File: src/main/java/com/bajdcc/LALR1/grammar/runtime/RuntimeTools.java
package com.bajdcc.LALR1.grammar.runtime;

import com.bajdcc.LALR1.grammar.runtime.RuntimeException.RuntimeError;
import com.bajdcc.LALR1.grammar.type.TokenTools;
import com.bajdcc.util.lexer.token.Token;

/**
 * 【运行时】运行时辅助类
 *
 * @author bajdcc
 */
public class RuntimeTools {

	public static boolean calcOp(RuntimeRegister reg, RuntimeInst inst,
	                             IRuntimeStack stk) throws RuntimeException {
		switch (inst) {
			case inot:
			case iinv:
			case iinc:
			case idec: {
				RuntimeObject obj = stk.load();
				if (obj == null) {
					throw new RuntimeException(RuntimeError.NULL_OPERATOR,
							reg.execId, "不允许空值运算");
				}
				if (obj.getType() == RuntimeObjectType.kNan) {
					stk.store(obj);
					return true;
				}
				Token tk = Token.createFromObject(obj.getObj());
				if (TokenTools.sin(TokenTools.ins2op(inst), tk)) {
					stk.store(new RuntimeObject(tk.object));
				} else {
					return false;
				}
			}
			break;
			case icl:
			case icg:
			case icle:
			case icge:
			case ice:
			case icne:
			case iadd:
			case iand:
			case iandl:
			case idiv:
			case imod:
			case imul:
			case ior:
			case iorl:
			case ishl:
			case ishr:
			case isub:
			case ixor: {
				RuntimeObject obj2 = stk.load();
				RuntimeObject obj = stk.load();
				if (obj.getType() == RuntimeObjectType.kNan) {
					stk.store(obj);
					return true;
				}
				if (obj2.getType() == RuntimeObjectType.kNan) {
					stk.store(obj2);
					return true;
				}
				Token tk = Token.createFromObject(obj.getObj());
				if (TokenTools.bin(TokenTools.ins2op(inst), tk,
						Token.createFromObject(obj2.getObj()))) {
					stk.store(new RuntimeObject(tk.object));
				} else {
					return false;
				}
			}
			break;
			default:
				break;
		}
		return true;
	}

	public static boolean calcJump(RuntimeRegister reg, RuntimeInst inst,
	                               IRuntimeStack stk) throws RuntimeException {
		switch (inst) {
			case ijmp:
				stk.opJump();
				break;
			case ijt:
				stk.opJumpBool(true);
				break;
			case ijf:
				stk.opJumpBool(false);
				break;
			case ijtx:
				stk.opJumpBoolRetain(true);
				break;
			case ijfx:
				stk.opJumpBoolRetain(false);
				break;
			case ijz:
				stk.opJumpZero(true);
				break;
			case ijnz:
				stk.opJumpZero(false);
				break;
			case ijyld:
				stk.opJumpYield();
				break;
			case ijnan:
				stk.opJumpNan();
				break;
			default:
				return false;
		}
		return true;
	}

	public static boolean calcData(RuntimeRegister reg, RuntimeInst inst,
	                               IRuntimeStack stk) throws Exception {
		switch (inst) {
			case ialloc:
				stk.opStoreDirect();
				break;
			case icall:
				stk.opCall();
				break;
			case ildfun:
				stk.opLoadFunc();
				break;
			case irefun:
				stk.opReloadFunc();
				break;
			case iload:
				stk.opLoad();
				break;
			case iloada:
				stk.opLoadArgs();
				break;
			case iloadv:
				stk.opLoadVar();
				break;
			case inop:
				break;
			case iopena:
				stk.opOpenFunc();
				break;
			case ipop:
				stk.pop();
				break;
			case ipush:
				stk.push();
				break;
			case ipusha:
				stk.opPushArgs();
				break;
			case ipushx:
				stk.opPushNull();
				break;
			case ipushz:
				stk.opPushZero();
				break;
			case ipushn:
				stk.opPushNan();
				break;
			case iret:
				stk.opReturn();
				break;
			case istore:
				stk.opStore();
				break;
			case icopy:
				stk.opStoreCopy();
				break;
			case iimp:
				stk.opImport();
				break;
			case iloadx:
				stk.opLoadExtern();
				break;
			case icallx:
				stk.opCallExtern(false);
				break;
			case ically:
				stk.opCallExtern(true);
				break;
			case iyldi:
				stk.opYield(true);
				break;
			case iyldo:
				stk.opYield(false);
				break;
			case iyldl:
				stk.opYieldSwitch(false);
				break;
			case iyldr:
				stk.opYieldSwitch(true);
				break;
			case iyldx:
				stk.opYieldDestroyContext();
				break;
			case iyldy:
				stk.opYieldCreateContext();
				break;
			case iscpi:
				stk.opScope(true);
				break;
			case iscpo:
				stk.opScope(false);
				break;
			case iarr:
				stk.opArr();
				break;
			case imap:
				stk.opMap();
				break;
			case iidx:
				stk.opIndex();
				break;
			case iidxa:
				stk.opIndexAssign();
				break;
			case itry:
				stk.opTry();
				break;
			case ithrow:
				stk.opThrow();
				break;
			default:
				return false;
		}
		return true;
	}

	public static String getYieldHash(int stackLevel, int funcLevel,
	                                  String pageName, int line) {
		return String.format("%d#%d#%s#%d", stackLevel, funcLevel, pageName,
				line);
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/runtime/RuntimeTools.kt
package com.bajdcc.LALR1.grammar.runtime

import com.bajdcc.LALR1.grammar.type.TokenTools
import com.bajdcc.util.lexer.token.Token

/**
 * 【运行时】运行时辅助类
 *
 * @author bajdcc
 */
object RuntimeTools {

    @Throws(RuntimeException::class)
    fun calcOp(inst: RuntimeInst, stk: IRuntimeStack): Boolean {
        when (inst) {
            RuntimeInst.inot, RuntimeInst.iinv, RuntimeInst.iinc, RuntimeInst.idec -> {
                val obj = stk.load()
                if (obj.type == RuntimeObjectType.kNan) {
                    stk.store(obj)
                    return true
                }
                val tk = Token.createFromObject(obj.obj!!)
                if (TokenTools.sin(TokenTools.ins2op(inst), tk)) {
                    stk.store(RuntimeObject(tk.obj))
                } else {
                    return false
                }
            }
            RuntimeInst.icl, RuntimeInst.icg, RuntimeInst.icle, RuntimeInst.icge, RuntimeInst.ice, RuntimeInst.icne, RuntimeInst.iadd, RuntimeInst.iand, RuntimeInst.iandl, RuntimeInst.idiv, RuntimeInst.imod, RuntimeInst.imul, RuntimeInst.ior, RuntimeInst.iorl, RuntimeInst.ishl, RuntimeInst.ishr, RuntimeInst.isub, RuntimeInst.ixor -> {
                val obj2 = stk.load()
                val obj = stk.load()
                if (obj.type == RuntimeObjectType.kNan) {
                    stk.store(obj)
                    return true
                }
                if (obj2.type == RuntimeObjectType.kNan) {
                    stk.store(obj2)
                    return true
                }
                val tk = Token.createFromObject(obj.obj!!)
                if (TokenTools.bin(TokenTools.ins2op(inst), tk,
                                Token.createFromObject(obj2.obj!!))) {
                    stk.store(RuntimeObject(tk.obj))
                } else {
                    return false
                }
            }
            else -> {
            }
        }
        return true
    }

    @Throws(RuntimeException::class)
    fun calcJump(inst: RuntimeInst, stk: IRuntimeStack): Boolean {
        when (inst) {
            RuntimeInst.ijmp -> stk.opJump()
            RuntimeInst.ijt -> stk.opJumpBool(true)
            RuntimeInst.ijf -> stk.opJumpBool(false)
            RuntimeInst.ijtx -> stk.opJumpBoolRetain(true)
            RuntimeInst.ijfx -> stk.opJumpBoolRetain(false)
            RuntimeInst.ijz -> stk.opJumpZero(true)
            RuntimeInst.ijnz -> stk.opJumpZero(false)
            RuntimeInst.ijyld -> stk.opJumpYield()
            RuntimeInst.ijnan -> stk.opJumpNan()
            else -> return false
        }
        return true
    }

    @Throws(Exception::class)
    fun calcData(inst: RuntimeInst, stk: IRuntimeStack): Boolean {
        when (inst) {
            RuntimeInst.ialloc -> stk.opStoreDirect()
            RuntimeInst.icall -> stk.opCall()
            RuntimeInst.ildfun -> stk.opLoadFunc()
            RuntimeInst.irefun -> stk.opReloadFunc()
            RuntimeInst.iload -> stk.opLoad()
            RuntimeInst.iloada -> stk.opLoadArgs()
            RuntimeInst.iloadv -> stk.opLoadVar()
            RuntimeInst.inop -> {
            }
            RuntimeInst.iopena -> stk.opOpenFunc()
            RuntimeInst.ipop -> stk.pop()
            RuntimeInst.ipush -> stk.push()
            RuntimeInst.ipusha -> stk.opPushArgs()
            RuntimeInst.ipushx -> stk.opPushNull()
            RuntimeInst.ipushz -> stk.opPushZero()
            RuntimeInst.ipushn -> stk.opPushNan()
            RuntimeInst.iret -> stk.opReturn()
            RuntimeInst.istore -> stk.opStore()
            RuntimeInst.icopy -> stk.opStoreCopy()
            RuntimeInst.iimp -> stk.opImport()
            RuntimeInst.iloadx -> stk.opLoadExtern()
            RuntimeInst.icallx -> stk.opCallExtern(false)
            RuntimeInst.ically -> stk.opCallExtern(true)
            RuntimeInst.iyldi -> stk.opYield(true)
            RuntimeInst.iyldo -> stk.opYield(false)
            RuntimeInst.iyldl -> stk.opYieldSwitch(false)
            RuntimeInst.iyldr -> stk.opYieldSwitch(true)
            RuntimeInst.iyldx -> stk.opYieldDestroyContext()
            RuntimeInst.iyldy -> stk.opYieldCreateContext()
            RuntimeInst.iscpi -> stk.opScope(true)
            RuntimeInst.iscpo -> stk.opScope(false)
            RuntimeInst.iarr -> stk.opArr()
            RuntimeInst.imap -> stk.opMap()
            RuntimeInst.iidx -> stk.opIndex()
            RuntimeInst.iidxa -> stk.opIndexAssign()
            RuntimeInst.itry -> stk.opTry()
            RuntimeInst.ithrow -> stk.opThrow()
            else -> return false
        }
        return true
    }

    fun getYieldHash(stackLevel: Int, funcLevel: Int,
                     pageName: String, line: Int): String {
        return String.format("%d#%d#%s#%d", stackLevel, funcLevel, pageName,
                line)
    }
}
